### DISCUSSION TITLE
404 when serving statics files with an express.js app

### DISCUSSION THREAD

**Discussion by leodev9911** [vercel/community#7359] [Help] [answered]
### Summary

I am encountering difficulties deploying static files generated from a React project on Vercel, using Express.js to handle API routes. Despite configuring routes in vercel.json to serve static files from the /public/ directory ({ "src": "/(.*)", "dest": "/public/$1" }), I consistently receive a 404 error when attempting to access these files. When i access to de https://vercel-app-example-api/api/ everithing is fine but in the '/' route i have  404 

Setup: My application utilizes Express.js for API handling (api/index.js), and the static files (e.g., index.html, style.css) are generated as part of a React project build process.
Observations: The deployment process completes without errors, as confirmed by Vercel logs. However, attempts to access static files result in 404 errors, suggesting an issue with how Vercel serves these files despite correct configuration.

I verified directory structures and file names to ensure they align with configured routes in vercel.json.
Checked Vercel deployment logs for any indications of errors or misconfigurations.
Attempted forced browser cache refresh (Ctrl + Shift + R) to rule out local caching issues.

This is my vercel.json
{
    "version": 2,
    "builds": [
      { "src": "api/index.js", "use": "@vercel/node" }
    ],
    "routes": [
      { "src": "/api/(.*)", "dest": "api/index.js" },
      { "src": "/(.*)", "dest": "/public/$1" }
    ]
}
  

### Example

https://fullstack-open-phonebook-api.vercel.app/

### Steps to Reproduce

![image](https://github.com/vercel/community/assets/125922076/4956e0f3-a27c-4a15-bdf3-6ad6a73a0253)

**Comment by amyegan** [comment, marked as answer]
Hey @leodev9911. It looks like your project uses the deprecated `builds` and `routes` config. [Node.js runtime will automatically be used for the runtime for JS files in the `/api` directory](https://vercel.com/docs/functions/runtimes/node-js) so you can remove the `builds` part. I also recommend switching [`routes`](https://vercel.com/docs/projects/project-configuration#routes) to [`rewrites`](https://vercel.com/docs/projects/project-configuration#rewrites). That isn't likely to solve the whole problem, but should make it a bit easier to debug without the extra config getting in your way.

There's also a guide available that walks through [using Express.js with Vercel](https://vercel.com/guides/using-express-with-vercel). This is a good [starter](https://github.com/vercel/examples/tree/main/solutions/express) to use if you're not already familiar with adapting Express sites to a serverless environment.

The config doesn't look quite right to me, but I don't know the full setup of your repo. If you can share a [minimal reproducible example](https://vercel.com/guides/creating-a-minimal-reproducible-example) I would be happy to take a closer look

**Reply by leodev9911** [reply]
Hi, than you so much for the answer. You can see the repo here [https://github.com/leodev9911/fullstack-open-phonebook-api](https://github.com/leodev9911/fullstack-open-phonebook-api). This is an app to learn Express.js from the Fullstack Open [course](https://fullstackopen.com/), at this point I tried with the guide that you provide me and it doesn't work. The deploy is good, there no errors in the log and when i go to /api/persons it work well, but in the / directory it give 404 not found when I'm expecting my react  app

**Reply by amyegan** [reply]
Thanks for sharing the repo with me. You're on the right track. 

It looks like there's a slight error in the `vercel.json` where `src` should be `source`. And you should either rename `dist` to `public` or override the [Output Directory](https://vercel.com/docs/deployments/configure-a-build#output-directory) to `dist` so the files will be available when deployed.

After fixing those two issues you should be able to deploy the update and see it working without a 404.

**Reply by leodev9911** [reply]
Thank you very much for you support🤗, It works!!!

[a comment by an automated account is not shown]
